libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,747 @@
+ * or `recvState`.
+ */
+static MQTTPublishState_t _calculateAckState( MQTTStateOperation_t opType,
+                                              MQTTPacketIDState_t idStatus,
+                                              MQTTPublishState_t sendState,
+                                              MQTTPublishState_t recvState );
+
+/**
+ * @brief Calculate a state when QoS is 1.
+ *
+ * @param[in] packetType PUBLISH or PUBACK.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
+ *
+ * @return The calculated state.
+ */
+static MQTTPublishState_t _calculateQoS1( MQTTPublishType_t packetType,
+                                          MQTTStateOperation_t opType,
+                                          MQTTPacketIDState_t idStatus );
+
+/**
+ * @brief Calculate a state when QoS is 2.
+ *
+ * @param[in] packetType PUBLISH, PUBREC, PUBREL, or PUBCOMP.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
+ *
+ * @return The calculated state.
+ */
+static MQTTPublishState_t _calculateQoS2( MQTTPublishType_t packetType,
+                                          MQTTStateOperation_t opType,
+                                          MQTTPacketIDState_t idStatus );
+
+/**
+ * @brief Test if a transition to new state is possible.
+ *
+ * @param[in] currentState The current state.
+ * @param[in] newState State to transition to.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] qos 0, 1, or 2.
+ *
+ * @note This function does not validate the current state, or the new state
+ * based on either the operation type or QoS. It assumes the new state is valid
+ * given the opType and QoS, which will be the case if calculated by
+ * _MQTT_CalculateState().
+ *
+ * @return `true` if transition is possible, else `false`
+ */
+static bool _validateTransition( MQTTPublishState_t currentState,
+                                 MQTTPublishState_t newState,
+                                 MQTTStateOperation_t opType,
+                                 MQTTQoS_t qos );
+
+/**
+ * @brief Test is a publish is outgoing or incoming.
+ *
+ * @param[in] packetType PUBLISH, PUBACK, PUBREC, PUBREL, or PUBCOMP.
+ * @param[in] opType Reserve, Send, or Receive.
+ *
+ * @return `true` if outgoing publish, else `false`
+ */
+static bool _isPublishOutgoing( MQTTPublishType_t packetType,
+                                MQTTStateOperation_t opType );
+
+/**
+ * @brief Reconcile a supplied QoS with one looked up from the record.
+ *
+ * @param[in] inputQoS 0, 1, 2, or INVALID.
+ * @param[in, out] pFoundQoS QoS retrieved from record.
+ *
+ * @return MQTTBadParameter if error; else MQTTSuccess.
+ */
+static MQTTStatus_t _resolveQoS( MQTTQoS_t inputQoS,
+                                 MQTTQoS_t * pFoundQoS );
+
+/**
+ * @brief Find a packet ID in the state record.
+ *
+ * @param[in] records State record array.
```

Comment:
The pattern of array+length is so common in our code that we should have a way to represent it without requiring 2 parameters.  This pattern is also extremely error-prone; bugs in which the wrong length gets associated with an array are very easy to write.  C++ addresses this with [`std::span`](https://en.cppreference.com/w/cpp/container/span), we should be able to come up with a simplified C version.

Reply:
I think the `sizeof(array)/sizeof(array_element)` approach is safe.